### PR TITLE
Remove max sample value

### PR DESCRIPTION
**Pull request by StephenButtolph**
## Why this should be merged

- We shouldn't rely on the std lib for exact consistency.
- We have a number of odd edge cases because we can't sample a number in the range (MaxInt64, MaxUint64].

## How this works

Extends the rng function to support generating all possible values in the range [0, MaxUint64]

## How this was tested

- Existing tests
- Manual edge case testing
- Fuzzing against the prior implementation for values [0, MaxInt64) (note, these must produce the exact same numbers when `n < MaxInt64` or else this could result in an unexpected change in the proposervm calculation.

### CODE REVIEW

File: utils/sampler/uniform_replacer.go [hunk 1]
```diff
@@ -82,7 +76,7 @@ func (s *uniformReplacer) Next() (uint64, error) {
 		return 0, errOutOfRange
 	}
 
-	draw := uint64(s.rng.Int63n(int64(s.length-s.drawsCount))) + s.drawsCount
+	draw := s.rng.Uint64n(s.length-1-s.drawsCount) + s.drawsCount
```
Comment (StephenButtolph) [inline utils/sampler/uniform_replacer.go:79]: Note the `-1` here now because `Uint64n` returns values in the range `[0, n]` whereas `Int63n` returned values in the range `[0, n)`

File: utils/sampler/uniform_resample.go [hunk 2]
```diff
@@ -70,7 +64,7 @@ func (s *uniformResample) Next() (uint64, error) {
 	}
 
 	for {
-		draw := uint64(s.rng.Int63n(int64(s.length)))
+		draw := s.rng.Uint64n(s.length - 1)
```
Comment (StephenButtolph) [inline utils/sampler/uniform_resample.go:67]: Note the `-1` here now because `Uint64n` returns values in the range `[0, n]` whereas `Int63n` returned values in the range `[0, n)`

File: utils/sampler/weighted_best.go [hunk 3]
```diff
@@ -41,15 +41,11 @@ func (s *weightedBest) Initialize(weights []uint64) error {
...
-	if totalWeight > math.MaxInt64 {
-		return errWeightsTooLarge
-	}
-
 	samples := []uint64(nil)
 	if totalWeight > 0 {
 		samples = make([]uint64, s.benchmarkIterations)
 		for i := range samples {
-			samples[i] = uint64(globalRNG.Int63n(int64(totalWeight)))
+			samples[i] = globalRNG.Uint64n(totalWeight - 1)
```
Comment (StephenButtolph) [inline utils/sampler/weighted_best.go:48]: Note the `-1` here now because `Uint64n` returns values in the range `[0, n]` whereas `Int63n` returned values in the range `[0, n)`

**Review by darioush — APPROVED** [review]
lgtm

File: utils/sampler/weighted_benchmark_test.go [hunk 4]
```diff
@@ -159,7 +153,7 @@ func WeightedPowBenchmarkInitializer(
 
 func WeightedSingletonBenchmarkInitializer(b *testing.B, s Weighted, size int) {
 	weights := make([]uint64, size)
-	weights[0] = uint64(math.MaxInt64 - size + 1)
+	weights[0] = math.MaxUint64 - uint64(size+1)
```
Comment (danlaine) [inline utils/sampler/weighted_benchmark_test.go:156]: Why'd this go from `- size + 1` to `- size - 1` ?

File: utils/sampler/weighted_benchmark_test.go [hunk 5]
```diff
@@ -119,14 +113,14 @@ func WeightedPowBenchmarkSampler(
...
-		_, _ = s.Sample(uint64(globalRNG.Int63n(int64(totalWeight))))
+		_, _ = s.Sample(globalRNG.Uint64n(totalWeight - 1))
 	}
 	return true
 }
 
 func WeightedSingletonBenchmarkSampler(b *testing.B, s Weighted, size int) bool {
 	weights := make([]uint64, size)
-	weights[0] = uint64(math.MaxInt64 - size + 1)
+	weights[0] = math.MaxUint64 - uint64(size+1)
```
Comment (danlaine) [inline utils/sampler/weighted_benchmark_test.go:123]: Why'd this go from `- size + 1` to `- size - 1`?

File: utils/sampler/weighted_benchmark_test.go [hunk 6]
```diff
@@ -119,14 +113,14 @@ func WeightedPowBenchmarkSampler(
 
 	b.ResetTimer()
 	for i := 0; i < b.N; i++ {
-		_, _ = s.Sample(uint64(globalRNG.Int63n(int64(totalWeight))))
+		_, _ = s.Sample(globalRNG.Uint64n(totalWeight - 1))
```
Comment (danlaine) [inline utils/sampler/weighted_benchmark_test.go:116]: Why does
```
s.Sample(uint64(globalRNG.Int63n(int64(totalWeight))))
```
become 
```
s.Sample(globalRNG.Uint64n(totalWeight - 1))
```
here but below the analogous call
```
s.Sample(uint64(globalRNG.Int63n(math.MaxInt64)))
```
becomes 
```
s.Sample(globalRNG.Uint64n(math.MaxUint64))
```
i.e. not with a `-1`?

File: utils/sampler/uniform_replacer.go (hunk 1 above)
Comment (danlaine) [reply utils/sampler/uniform_replacer.go:79]: Would it be simpler to just have `Uint64n` return a number in `[0, n)`?

Comment (patrick-ogrady) [reply utils/sampler/uniform_replacer.go:79]: Wondering if we should distinguish between these 2 random modes? Seems confusing that `*n` function now behaves differently (especially if it is different than the golang rand pkg: https://pkg.go.dev/math/rand#Int31n)?

File: utils/sampler/rand.go [hunk 7]
```diff
@@ -4,58 +4,92 @@
...
+		return v
+
+	// n is less than MaxUint64/2 so we generate a number in the range
+	// [0, k*(n+1)) where k is the largest integer such that k*(n+1) is less
+	// than or equal to MaxUint64/2. We can't easily find k such that k*(n+1) is
+	// less than or equal to MaxUint64 because the calculation would overflow.
+	//
+	// ref: https://github.com/golang/go/blob/ce10e9d84574112b224eae88dc4e0f43710808de/src/math/rand/rand.go#L127-L132
+	default:
+		max := (1 << 63) - 1 - (1<<63)%(n+1)
```
Comment (patrick-ogrady) [inline utils/sampler/rand.go:73]: Can we elaborate on this a bit or continue to call `(1 << 63) - 1` `int63mask`

Comment (StephenButtolph) [reply utils/sampler/rand.go:73]: This line is almost exactly taken from:  https://github.com/golang/go/blob/ce10e9d84574112b224eae88dc4e0f43710808de/src/math/rand/rand.go#L127-L132 and I didn't want to modify it more than needed to make understanding the diff easier.

Comment (patrick-ogrady) [reply utils/sampler/rand.go:73]: Sorry, I'm an idiot and missed the comment

**Review by patrick-ogrady — COMMENTED** [review]
This looks correct to me and all "novel" code is pulled from `math/rand`.

File: utils/sampler/uniform_replacer.go (hunk 1 above)
Comment (StephenButtolph) [reply utils/sampler/uniform_replacer.go:79]: > Would it be simpler to just have Uint64n return a number in [0, n)?

This would mean that we would need to panic if `n == 0` which I was trying to avoid.

File: utils/sampler/weighted_benchmark_test.go (hunk 4 above)
Comment (StephenButtolph) [reply utils/sampler/weighted_benchmark_test.go:156]: :eagle: :eye:

File: utils/sampler/weighted_benchmark_test.go (hunk 5 above)
Comment (StephenButtolph) [reply utils/sampler/weighted_benchmark_test.go:123]: :eagle: :eye:

File: utils/sampler/weighted_benchmark_test.go (hunk 6 above)
Comment (StephenButtolph) [reply utils/sampler/weighted_benchmark_test.go]: :eagle: :eye:

**Review by danlaine — APPROVED** [review]

File: utils/sampler/rand.go [hunk 8]
```diff
@@ -4,58 +4,92 @@
...
+var globalRNG = newRNG()
 
-func newRNG() rng {
+func newRNG() *rng {
 	source := prng.NewMT19937()
 	source.Seed(uint64(time.Now().UnixNano()))
-	// We don't use a cryptographically secure source of randomness here, as
-	// there's no need to ensure a truly random sampling.
-	return rand.New(&syncSource{rng: source}) // #nosec G404
+	return &rng{rng: source}
```
Comment (dnkolegov) [inline utils/sampler/rand.go:19]: Removing the comment suggests you will use this PRNG as a cryptographically secure one. I recommend leaving the comment here. It will also raise questions in future audits.

File: utils/sampler/uniform_replacer.go [hunk 9]
```diff
@@ -30,23 +28,19 @@ func (m defaultMap) get(key uint64, defaultVal uint64) uint64 {
...
+func (s *uniformReplacer) Initialize(length uint64) {
 	s.rng = globalRNG
 	s.seededRNG = newRNG()
 	s.length = length
 	s.drawn = make(defaultMap)
 	s.drawsCount = 0
-	return nil
 }
 
 func (s *uniformReplacer) Sample(count int) ([]uint64, error) {
```
Comment (dnkolegov) [inline utils/sampler/uniform_replacer.go:46]: This will panic if `count = -1`
```
s := uniformReplacer{}
s.Initialize(1)
s.Sample(-1)
```

File: utils/sampler/uniform_resample.go [hunk 10]
```diff
@@ -19,21 +17,17 @@ import (
...
-	}
+func (s *uniformResample) Initialize(length uint64) {
 	s.rng = globalRNG
 	s.seededRNG = newRNG()
 	s.length = length
 	s.drawn.Clear()
-	return nil
 }
 
 func (s *uniformResample) Sample(count int) ([]uint64, error) {
```
Comment (dnkolegov) [inline utils/sampler/uniform_resample.go:33]: Probably, this will function also panic if `count` is -1

**Review by dnkolegov — CHANGES_REQUESTED** [review]

File: utils/sampler/rand.go (hunk 8 above)
Comment (StephenButtolph) [reply utils/sampler/rand.go:21]: ok

File: utils/sampler/uniform_replacer.go (hunk 9 above)
Comment (StephenButtolph) [reply utils/sampler/uniform_replacer.go:46]: I feel like panicking is pretty reasonable when requesting a negative number of values

File: utils/sampler/uniform_replacer.go (hunk 1 above)
Comment (StephenButtolph) [reply utils/sampler/uniform_replacer.go]: Renamed `Uint64n` to `Uint64Inclusive` to be very clear
